x-pack/plugins/observability_ai_assistant/public/functions/recall.ts

```diff
@@ -72,28 +76,216 @@ export function registerRecallFunction({
+  queries: string[];
+  service: ObservabilityAIAssistantService;
+  connectorId: string;
+  signal: AbortSignal;
+}) {
+  const systemMessageExtension =
+    dedent(`You have the function called score available to help you inform the user about how relevant you think a given document is to the conversation.
+    Please give a score between 1 and 7, fractions are allowed.
+    A higher score means it is more relevant.`);
+  const extendedSystemMessage = {
+    ...systemMessage,
+    message: {
+      ...systemMessage.message,
+      content: `${systemMessage.message.content}\n\n${systemMessageExtension}`,
+    },
+  };
+
+  const userMessageOrQueries =
+    userMessage && userMessage.message.content ? userMessage.message.content : queries.join(',');
+
+  const newUserMessageContent =
+    dedent(`Given the question "${userMessageOrQueries}", can you give me a score for how relevant the following documents are?
+
+  ${JSON.stringify(suggestions, null, 2)}`);
+
+  const newUserMessage: Message = {
+    '@timestamp': new Date().toISOString(),
+    message: {
+      role: MessageRole.User,
+      content: newUserMessageContent,
+    },
+  };
+
+  const scoreFunction = {
+    name: 'score',
+    description:
+      'Use this function to score documents based on how relevant they are to the conversation.',
+    parameters: {
+      type: 'object',
+      additionalProperties: false,
+      properties: {
+        scores: {
+          description: 'The document IDs and their scores',
+          type: 'array',
+          items: {
+            type: 'object',
+            additionalProperties: false,
+            properties: {
+              id: {
+                description: 'The ID of the document',
+                type: 'string',
+              },
+              score: {
+                description: 'The score for the document',
+                type: 'number',
+              },
             },
           },
-          signal,
-        })
-        .then((response): { content: Serializable } => ({
-          content: response.entries.map((entry) =>
-            omit(entry, 'labels', 'score', 'is_correction')
-          ) as unknown as Serializable,
-        }));
-    }
+        },
+      },
+      required: ['score'],
+    } as const,
+    contexts: ['core'],
+  };
+
+  const response = (await service.callApi('POST /internal/observability_ai_assistant/chat', {
+    params: {
+      query: {
+        stream: false,
+      },
+      body: {
+        connectorId,
+        messages: [extendedSystemMessage, newUserMessage],
+        functions: [scoreFunction],
+        functionCall: 'score',
+      },
+    },
+    signal,
+  })) as CreateChatCompletionResponse;
+
+  const scoreFunctionRequest = decodeOrThrow(scoreFunctionRequestRt)(response);
+  const { scores } = decodeOrThrow(jsonRt.pipe(scoreFunctionArgumentsRt))(
+    scoreFunctionRequest.choices[0].message.function_call.arguments
   );
+
+  if (scores.length === 0) {
+    return [];
+  }
```

Comment:
Not sure if we should throw here instead, since it means the LLM didn't score any document.
Perhaps it would be more correct to verify that all the IDs in the request also appear in the response?

Reply:
@dgieselaar Thoughts on this?

Reply:
No need to throw I think. I think the LLM could very well forgot to score documents or hallucinate entire documents, but we shouldn't throw in that scenario. Just filter out the ones that do not exist.